src/com/facebook/buck/ide/intellij/ModuleBuildContext.java

```diff
@@ -153,23 +155,43 @@ public void addSourceFolder(IjFolder folder) {
     Path path = folder.getPath();
     IjFolder otherFolder = sourceFoldersMergeMap.get(path);
     if (otherFolder != null) {
-      folder = mergeAllowingTestToBePromotedToSource(folder, otherFolder);
+      folder = mergePromotingToSourceIfDifferent(folder, otherFolder);
     }
     sourceFoldersMergeMap.put(path, folder);
   }
 
-  private IjFolder mergeAllowingTestToBePromotedToSource(IjFolder from, IjFolder to) {
-    if ((from instanceof TestFolder && to instanceof SourceFolder)
-        || (to instanceof TestFolder && from instanceof SourceFolder)) {
-      return new SourceFolder(
-          to.getPath(),
-          from.getWantsPackagePrefix() || to.getWantsPackagePrefix(),
-          IjFolder.combineInputs(from, to));
+  /**
+   * Merges the two folders according to the following rules: - Folders of the same type merge
+   * normally. - A resource folder and a test resource folder should become a regular resource
+   * folder. - A folder that contains both regular and test sources should become a source folder. -
+   * A folder that contains any kind of source and any kind of resource will not work correctly if
+   * the resources_root is different from the src_root, since buck allows individual files to be
+   * marked as sources/resources, but IntelliJ's granularity is only at the folder level. Thus, we
+   * try to handle this by promoting these to source folders, but there isn't really a good way
+   * around it.
+   */
+  private IjFolder mergePromotingToSourceIfDifferent(IjFolder from, IjFolder to) {
+    if (from.getClass().equals(to.getClass())) {
+      return from.merge(to);
     }
 
-    Preconditions.checkArgument(from.getClass() == to.getClass());
+    // Both resource folders but one is a test resource and one is a regular resource.
+    if ((from.getClass().equals(JavaResourceFolder.class)
+            && to.getClass().equals(JavaTestResourceFolder.class))
+        || (to.getClass().equals(JavaResourceFolder.class)
+            && from.getClass().equals(JavaTestResourceFolder.class))) {
+      Path resourcesRoot =
+          (to instanceof JavaResourceFolder)
+              ? ((JavaResourceFolder) to).getResourcesRoot()
+              : ((JavaTestResourceFolder) to).getResourcesRoot();
+      return new JavaResourceFolder(to.getPath(), IjFolder.combineInputs(from, to), resourcesRoot);
+    }
 
-    return from.merge(to);
+    // If we're not sure what to do otherwise, use SourceFolder.
+    return new SourceFolder(
```

Comment:
Changed, though actually I think regardless of whether it's set to SourceFolder or TestFolder in this case, it's not going to work correctly, but TestFolder is probably a better attempt in that case